### ISSUE TITLE
Create new top-level section "Foundations" for housing non-CS prerequisites?

### ISSUE THREAD

**Issue by hanjiexi** [ossu/computer-science#531]
As I described in #438, I am thinking about pulling most non-CS topics out of Intro/Core/Advanced CS and putting them in a separate section called "Foundations". Courses from this section would only be required when someone doesn't have the relevant background, and the curriculum would still begin with Intro CS. Students would only take what they need from it, when they need it. The prerequisites section of each course would then contain a hyperlink to the relevant section in Foundations.

Rough outline of how that would play out, assuming we were to adopt other changes I've suggested recently:

# Foundations

Complete only those sections as directed by the prerequisites of a CS course you're about to take.

## K-8 math
Select the appropriate grade level from [Khan Academy Math](https://www.khanacademy.org/math).

## High school math
- Depends on: K-8 Math
- [Khan Academy: High school math](https://www.khanacademy.org/math/math-1-2-3) (all sections)
- For extra practice / refreshers: [ASU College Algebra](https://www.edx.org/course/college-algebra-problem-solving-asux-mat117x) and [ASU Precalculus](https://www.edx.org/course/precalculus-asux-mat170x)
- Getting help: [/r/learnmath](https://www.reddit.com/r/learnmath/)

## Single-variable calculus
- Depends on: High school math
- Full courses: MIT Calculus 1A 1B 1C
- For extra practice / refreshers: [Expii Calculus 1](https://www.expii.com/t/calculus-1)
- Getting help: [/r/learnmath](https://www.reddit.com/r/learnmath/)

## Multivariable calculus
- Depends on: Single-variable calculus
- Full course: [Multivariable Calculus](https://ocw.mit.edu/courses/mathematics/18-02sc-multivariable-calculus-fall-2010/index.htm)
- For extra practice / refreshers: [Khan Academy Multivariable Calculus ](https://www.khanacademy.org/math/multivariable-calculus)
- Getting help: [/r/learnmath](https://www.reddit.com/r/learnmath/)

## Probability
- Depends on: Multivariable calculus
- Full course: [Introduction to Probability](https://www.edx.org/course/introduction-probability-science-mitx-6-041x-2)
- For extra practice / refreshers: [Khan Academy Statistics and probability](https://www.khanacademy.org/math/statistics-probability)

## Electricity
- Depends on: Single-variable calculus
- Full course sequence:
    1. [Introduction to Mechanics, Part 1](https://www.edx.org/course/introduction-mechanics-part-1-ricex-phys-101-1x)
    2. [Electricity and Magnetism, Part 1](https://www.edx.org/course/electricity-magnetism-part-1-ricex-phys102-1x-1)
    3. [Electricity and Magnetism, Part 2](https://www.edx.org/course/electricity-magnetism-part-2-ricex-phys102-2x-0)
- For extra practice / review: [Expii Mechanics](https://www.expii.com/t/mechanics-486) and [Expii Electricity, Magenetism, and Optics](https://www.expii.com/t/electricity-magnetism-and-optics-1620)

## Modern technology
- Stanford CS101 (see the discussion in #530)

# Intro CS
- Depends on: Foundations#Modern technology
- Python prep for those who have never programmed in their life: [Computing in Python](https://www.edx.org/professional-certificate/introduction-to-computing-in-python) 1-3
- MIT Intro to CS and Programming

# Core CS

### Linear algebra
- Depends on: Foundations#High school math
- [Textbook supplement](https://www.math.ucdavis.edu/~anne/linear_algebra/index.html)
- Course sequence:
    1. Essence of Linear Algebra
    2. Linear Algebra Foundations to Frontiers

### Discrete mathematics
- Introductory reading: [Notes on discrete math](http://www.cs.elte.hu/~lovasz/dmbook.ps)
- Prep for those who need more exercises: [Discrete Math Specialization](https://www.coursera.org/specializations/discrete-mathematics)
- Full course: MIT Math for CS courseware and [textbook](https://courses.csail.mit.edu/6.042/spring17/mcs.pdf)

(etc.)

# Advanced CS

## Advanced machine systems
- Depends on: Foundations#Electricity
- Full course sequence: MIT Computation Structures

(etc.)

**Comment by waciumawanjohi** [comment]
I like this idea in general. A future goal would be to have an assessment that would let students know if they can skip a course. (CLEP has an Algebra, Precal and Calculus class) If we are listing electricity, we should list high school physics.

On math, I suspect linear algebra should sit alongside calculus and probability in Foundations math, with only discrete math in the required math courses.

Python programming should receive more (not less) emphasis than Stanford's CS101.

We'll help students evaluate if they should take a Foundations course by not only listing its pre-reqs, but also what courses it prepares students for.

**Comment by Alaharon123** [comment]
Whether to put Calculus and up in Foundations or Core depends on whether the goal is for Core to be everything in a cs degree besides for gened or if it's to only include cs specific courses. If it's the first, most cs degrees assume you already know precal and start with Calculus so that would be where ossu starts from. If it's the second then Calculus would be in Foundations.

Sidenote: I don't know how practical this is, but you could consider making something like University of Buffalo's [flowcharts] (https://engineering.buffalo.edu/home/academics/undergrad/advisement/flowsheets.html?plan=CSEBS-2018). They let you mouse over courses to see prerequisites and courses that use it as a prereq.

**Comment by chinguelessi** [comment]
Hello!

I just want to say that I like this idea a lot. I was looking for guidance on the best resource to learn/refresh the necessary high school math without knowing what is relevant and what is not. Much beacause I was not sure what the American curriculum demands. I am following this site, https://www.mathsisfun.com/links/curriculum-high-school-algebra.html, but then I am like, is it enough? Is it too much? So, having some guidance is really useful.

Also, Khan Academy is great, but bandwidth is really expensive where I live, so whenever I can, I avoid videos.

Thank you!

**Comment by CamLV** [comment]
I love the idea of the Foundations page. 
Back in the 70's, I studied Engineering at Georgia Tech. Dropped out. In the 80's, I studied Computer Science at Rutgers. Dropped out again. At both, calculus was a required course. I think the Core Math section should section should include Single Variable Calculus and Multivariable Calculus to reflect the both structure of a degree program, the importance of these courses to our studies and the effort needed to understand and pass them. Personally, I hated Calculus but loved Linear Algebra. 
Perhaps we could link each course to a page or pop-up that lists things like alternatives, testbooks, sites for practice and places to go for help to keep the main page and the foundations page cleaner?

**Comment by hanjiexi** [comment]
Thanks for everyone's feedback!

@waciumawanjohi 
> On math, I suspect linear algebra should sit alongside calculus and probability in Foundations math, with only discrete math in the required math courses.

It should, yes, but currently we use LAFF which uses a lot of programming exercises written in a style similar to how one writes linear algebra programs in C, etc. Linear algebra could go in Foundations if we used a more abstract resource with no to little programming ([Strang's materials](https://ocw.mit.edu/courses/mathematics/18-06-linear-algebra-spring-2010/video-lectures/) probably are like this).

> Python programming should receive more (not less) emphasis than Stanford's CS101.

I'm not sure what you're saying here. Are you suggesting we don't use Stanford CS101? Or add more Python courses? What's your rationale? (Not challenging, just asking for clarity)

> We'll help students evaluate if they should take a Foundations course by not only listing its pre-reqs, but also what courses it prepares students for.

Reverse dependencies are a good idea!

@Alaharon123 
> Whether to put Calculus and up in Foundations or Core depends on whether the goal is for Core to be everything in a cs degree besides for gened or if it's to only include cs specific courses. If it's the first, most cs degrees assume you already know precal and start with Calculus so that would be where ossu starts from. If it's the second then Calculus would be in Foundations.

@CamLV 
> I think the Core Math section should section should include Single Variable Calculus and Multivariable Calculus to reflect the both structure of a degree program, the importance of these courses to our studies and the effort needed to understand and pass them.

Thanks for your input. I understand your perspectives, but I want to point out that even though I'm proposing to move Calculus, et al., off the Core CS *page*, it's still required for everyone who hasn't already taken it and would be clearly marked as such in the relevant sections.

Regarding geneds: I'm trying to solve a short-term problem specific to OSSU CS, but general education requirements (as commonly, *and rightly*, required by CS standards committees) isn't just an OSSU CS problem; it's an OSSU problem. Geneds are, by definition, required of everyone at a university, so it should be solved at the OSSU level. Maybe someone else can step up and create a general education curriculum that we could then refer to.

@chinguelessi 
> Also, Khan Academy is great, but bandwidth is really expensive where I live, so whenever I can, I avoid videos.

Not having an answer for this situation is an oversight of OSSU and I apologize for that! But as you can probably tell from above, I want to start including text-based resources alongside videos for people who prefer (or need) this format.

**Comment by waciumawanjohi** [comment]
> > On math, I suspect linear algebra should sit alongside calculus and probability in Foundations math, with only discrete math in the required math courses.

> It should, yes, but currently we use LAFF which uses a lot of programming exercises written in a style similar to how one writes linear algebra programs in C

If Linear Algebra is in the core curriculum only because the only acceptable course uses programming, we should ensure that students know which courses rely on LinAlg and that their selected path through the curriculum may not require the course.

> I'm not sure what you're saying here. 

The wording sounded to me like CS101 was being emphasized more than CS1301. Not a big deal.

**Comment by Alaharon123** [comment]
@chinguelessi https://courses.edx.org/courses/course-v1:ASUx+MAT117x+1T2016/course/ is really good especially for returning students. It tests your knowledge and then only teaches you what you need to know. It also checks every so often to make sure you've mastered topics so you don't get left behind with something you don't _really_ understand. It starts off with a knowledge check. If you start after that with less than 15% in your pie, then you're not ready for college algebra yet. If you start with more than 85%, you can skip to [Pre-Calculus](https://courses.edx.org/courses/course-v1:ASUx+MAT170x+2T2017/course/). It's all text and pictures so it should take less bandwidth than videos. Of course, this is only relevant if you're ready for College Algebra. If you get less than 15%, then you still have the same problem. I've heard the OpenStax textbooks, which start all the way from pre-algebra, are good, but I haven't used them so ymmv. You could probably take the quizzes on Khan Academy to figure out where you're holding and download a pdf (free) or buy a physical copy (paid) of the appropriate textbook.

@hanjiexi gened for ossu is definitely doable. Using the SUNY requirements as an example since I'm familiar with them, Mathematics is already covered, there are plenty of online courses for learning a language, social sciences can be some course on Computers in Society (that's what it is in my college), Arts can be an introductory html/css course (I'm taking that right now), I think there's such a thing as an online physics lab course. The hardest part, other than actually choosing the courses (requires knowledge of courses from a variety of fields so that's actually pretty hard), would probably be finding a college level online english course. Not sure how that would even work.

**Comment by hanjiexi** [comment]
@waciumawanjohi 
> If Linear Algebra is in the core curriculum only because the only acceptable course uses programming, we should ensure that students know which courses rely on LinAlg and that their selected path through the curriculum may not require the course.

That's a good point. It will be very inconvenient to have a section in Core Math that isn't actually needed except in some very specific domains. From a brief glance and trying to keep all the proposed changes in my head, linear algebra is only needed for one course in Advanced CS (LAFF Programming for Correctness) and several courses that I proposed to move to Domains (e.g. graphics, machine learning).

I have some invariants I hope to maintain for the top-level sections, but without mapping all the proposed changes I don't know yet how this will play out:
1. No top-level section should contain forward-dependencies.
    - Example 1: a course in Foundations can depend on another Foundations course, but it shouldn't depend on material from Core CS or above.
    - Example 2: a course in Core CS can depend on material from Core CS and Foundations, but not on material in Advanced CS.
2. No top-level section *except Domains* (our endpoint) should contain unused dependencies. If there is no way to trace the reverse dependencies of material in Foundations, Core, and Advanced all the way to some Domain / application area, then the material belongs in Extras.
    - Example: Logic programming (fulfilled via Prolog) currently isn't used anywhere beyond Core Programming. It should be moved to Extras, unless we add a Domain that uses it or depends on material that uses it.
3. Every section in Advanced CS should have a corresponding section in Core CS. If there is no way to justify adding a corresponding section in Core CS, then such section belongs in Domains.
    - This is relevant to our discussion in #528.
4. There needs to be at least one section of Advanced CS that only depends on Foundations or Core CS, not on other sections of Advanced CS.
    - For this to be true of all sections is a "nice to have" but it might be too hard to maintain.

My basic idea is to model a four-year undergraduate CS degree, where this progression is possible:
- Year 1 begins with Intro CS, and the remainder of the time is spent fulfilling prerequisites in Foundations (possibly in geneds if we have them), with any remaining time on early Core CS coursework.
- Year 2 is for finish the rest of Core CS, having already done everything needed in Foundations.
- Year 3 is spent on one selected section of Advanced CS, which would have to be one that doesn't depend on any other section in Advanced CS (see invariant 4).
- Year 4 is spent either on one Domain or a second section of Advanced CS.

**Comment by hanjiexi** [comment]
So with the above design (for which I, like everything else, hope to get feedback), that means Linear Algebra would likely end up in Foundations or Advanced CS. If it goes in Foundations, the resource we use can't depend on advanced programming knowledge. If it goes in Advanced Math, we would need to show that Advanced Math is a viable subject in and of itself for year 3 or 4, and that such material is depended on by at least one Domain.

(I might be overreaching on the amount of work that is viable to do before v9 lands. I want v9 to arrive by the end of the year.)

**Comment by hanjiexi** [comment]
I have used our wiki to create a draft for what the new Foundations page could look like: https://github.com/ossu/computer-science/wiki/Foundations-draft

Notes:
- No narrative written yet for any individual sections (and usually should be self-explanatory)
- The Prerequisites and "Needed by" sections don't yet have hyperlinks, but they will when everything is finalized
- The "Use cases" headings are important in the table where I list Resources:
  + When it says `learn`, that means we are recommending this resource as the one to use when learning it for the first time.
  + If it says `review`, this means that you should use it if you have already learned the material but need to review it, perhaps because you learned it a long time ago.
  + If it says `practice`, this means it's a resource meant for when you have already learned it but need more exercises.
  + (Currently, I always list the same resource as both `review` and `practice`, but we should pick better individual resources for these two activities.)
  + If it says `prepare`, this means the material isn't dense enough to really learn from (it may just be video(s)), but it helps to prepare the student for whatever comes next. This is the case with Essence of Linear Algebra.
  + If it says `help`, that means this is the resource we're recommending for those who need help with the material.
- I originally didn't have a section labeled Physics, but did have one named Electricity. @waciumawanjohi recommended I make a separate physics section, but I looked closer at the courses under Electricity and found that those are the very first courses a student would take in physics and don't appear to assume any prior physics knowledge. So I basically just renamed "Electricity" to "Physics". I wish we could find a course like "Physics for CS" as I'm sure much of the material is not relevant.
- Of note about Linear Algebra: currently I'm listing two tracks, one is LAFF and the other is Gilbert Strang's course. Although LAFF has programming assignments, they are optional, so I think it's fine to include it in Foundations (and they don't assume any prior programming knowledge anyway). Strang's course is listed as a more challenging and rigorous alternative — I didn't want to leave it out since it is probably the very best course on the subject — but it requires MATLAB and Multivariable Calculus, so not an option for most students.

I will proceed to format the rest of the curriculum pages in this way unless there is anything seriously wrong with it.

**Comment by Alaharon123** [comment]
Looks good. Should probably also have a flowchart to give an overarching view of everything.

**Comment by Alaharon123** [comment]
Looks good, but it's unclear why Applied Linear Algebra is here rather than in extras/readings. Why would one want to reference this incomplete text?

**Comment by hanjiexi** [comment]
I want to ensure that there are text-based resources listed alongside video-based resources for people whose bandwidth is expensive. I don't know how this will play out yet, but I'm thinking that the separate extras/readings page might be removed in favor of keeping all high-quality materials grouped under the relevant topic, with explicit recommendations given about which resources are worth using under which conditions.

Applied Linear Algebra, incomplete though it may be, is geared towards beginners with only a high school math background. The other free text, by Jim Hefferson, assumes college level calculus just like the Strang course. So both groups of resources are grouped according to their prerequisites. I do probably need to add a note about the incompleteness of Applied Linear Algebra compared to the course it sits alongside.

At some point I also need to figure out just how much of linear algebra exactly is required for which later courses. It may be that some courses need a lot more and some need much less, so the tracks could be re-organized later according to their reverse dependencies. But that's not going to happen anytime soon as I don't have the time.

**Comment by Alaharon123** [comment]
Ok I made a mistake. I saw Immersive Linear Algebra in the text and then Applied Linear Algebra in the table and didn't realize they weren't the same thing. What I meant to ask was why Immersive Linear Algebra is put there if that's something for OSSU to keep an eye on rather than for students to keep an eye on. However, I am now realizing that this is because you are using the wiki to have notes for yourself and are probably planning on transferring this all to the readme when 9.0 is ready. 

I see that you added a note afterwards stating that Applied Linear Algebra does not cover everything it should. If that's because of my comment, I was mixing it up with Immersive Linear Algebra (if that's a thing anyway then I guess I'm glad that my mistaken comment led to something positive). Sorry for the confusion. The original text was clear, just Applied and Immersive look similar enough that I had a brain fart.
